README.md
[![NPM version](https://badge.fury.io/js/cdk-aurora-globaldatabase.svg)](https://badge.fury.io/js/cdk-aurora-globaldatabase)
[![PyPI version](https://badge.fury.io/py/cdk-aurora-globaldatabase.svg)](https://badge.fury.io/py/cdk-aurora-globaldatabase)
![Release](https://github.com/neilkuan/cdk-aurora-globaldatabase/workflows/Release/badge.svg)

![Downloads](https://img.shields.io/badge/-DOWNLOADS:-brightgreen?color=gray)
![npm](https://img.shields.io/npm/dt/cdk-aurora-globaldatabase?label=npm&color=orange)
![PyPI](https://img.shields.io/pypi/dm/cdk-aurora-globaldatabase?label=pypi&color=blue)

# cdk-aurora-globaldatabase
`cdk-aurora-globaldatabase` is an AWS CDK construct library that allows you to create [Amazon Aurora Global Databases](https://aws.amazon.com/rds/aurora/global-database/) with AWS CDK in Typescript or Python.

# Why
**Amazon Aurora Global Databases** is designed for multi-regional Amazon Aurora Database clusters that span across different AWS regions. Due to the lack of native cloudformation support, it has been very challenging to build with cloudformation or AWS CDK with the upstream `aws-rds` construct.

`cdk-aurora-globaldatabase` aims to offload the heavy-lifting and helps you provision and deploy cross-regional **Amazon Aurora Global Databases** simply with just a few CDK statements.

## Now Try It !!!
# Sample for Mysql 
```ts
import { GolbalAuroraRDSMaster, InstanceTypeEnum, GolbalAuroraRDSSlaveInfra } from './index';
import { App, Stack, CfnOutput } from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
// new app .
const mockApp = new App();

// setting two region env config .
const envSingapro  = { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'ap-southeast-1' };
const envTokyo = { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'ap-northeast-1' };

// create stack main .
const stackM = new Stack(mockApp, 'testing-stackM',{env: envTokyo});
const vpcPublic = new ec2.Vpc(stackM,'defaultVpc',{
  natGateways: 0,
  maxAzs: 3,
  subnetConfiguration: [{
    cidrMask: 26,
    name: 'masterVPC2',
    subnetType: ec2.SubnetType.PUBLIC,
  }],
});
const  globaldbM = new GolbalAuroraRDSMaster(stackM, 'golbalAuroraRDSMaster',{
  instanceType: InstanceTypeEnum.R5_LARGE,
  vpc: vpcPublic,
  rdsPassword: '1qaz2wsx',
});
globaldbM.rdsCluster.connections.allowDefaultPortFrom(ec2.Peer.ipv4(`${process.env.MYIP}/32`))

// create stack slave infra or you can give your subnet group.
const stackS = new Stack(mockApp, 'testing-stackS',{env: envSingapro});
const vpcPublic2 = new ec2.Vpc(stackS,'defaultVpc2',{
  natGateways: 0,
  maxAzs: 3,
  subnetConfiguration: [{
    cidrMask: 26,
    name: 'secondVPC2',
    subnetType: ec2.SubnetType.PUBLIC,
  }],
});
const globaldbS = new GolbalAuroraRDSSlaveInfra(stackS, 'slaveregion',{vpc: vpcPublic2,subnetType:ec2.SubnetType.PUBLIC });

// so we need to wait stack slave created first .
stackM.addDependency(stackS)


new CfnOutput(stackM, 'password', { value: globaldbM.rdsPassword });
// add second region cluster
globaldbM.addRegionalCluster(stackM,'addregionalrds',{
  region: 'ap-southeast-1',
  dbSubnetGroupName: globaldbS.dbSubnetGroup.dbSubnetGroupName,
});
```
![like this ](./image/Mysql-cluster.jpg)
# Sample for Postgres

```ts
import { GolbalAuroraRDSMaster, InstanceTypeEnum, GolbalAuroraRDSSlaveInfra } from './index';
import { App, Stack, CfnOutput } from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as _rds from '@aws-cdk/aws-rds';

const mockApp = new App();
const envSingapro  = { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'ap-southeast-1' };
const envTokyo = { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'ap-northeast-1' };

const stackM = new Stack(mockApp, 'testing-stackM',{env: envTokyo});
const vpcPublic = new ec2.Vpc(stackM,'defaultVpc',{
  natGateways: 0,
  maxAzs: 3,
  subnetConfiguration: [{
    cidrMask: 26,
    name: 'masterVPC2',
    subnetType: ec2.SubnetType.PUBLIC,
  }],
});

// Note if you use postgres , need to give the same value in engineVersion and  dbClusterpPG's engine .
const globaldbM = new GolbalAuroraRDSMaster(stackM, 'golbalAuroraRDSMaster',{
  instanceType: InstanceTypeEnum.R5_LARGE,
  vpc: vpcPublic,
  rdsPassword: '1qaz2wsx',
  engineVersion: _rds.DatabaseClusterEngine.auroraPostgres({
    version: _rds.AuroraPostgresEngineVersion.VER_11_7}),
  dbClusterpPG: new _rds.ParameterGroup(stackM, 'dbClusterparametergroup', {
    engine: _rds.DatabaseClusterEngine.auroraPostgres({
      version: _rds.AuroraPostgresEngineVersion.VER_11_7,
    }),
    parameters: {
      'rds.force_ssl': '1',
      'rds.log_retention_period': '10080',
      'auto_explain.log_min_duration': '5000',
      'auto_explain.log_verbose': '1',
      'timezone': 'UTC+8',
      'shared_preload_libraries': 'auto_explain,pg_stat_statements,pg_hint_plan,pgaudit',
      'log_connections': '1',
      'log_statement': 'ddl',
      'log_disconnections': '1',
      'log_lock_waits': '1',
      'log_min_duration_statement': '5000',
      'log_rotation_age': '1440',
      'log_rotation_size': '102400',
      'random_page_cost': '1',
      'track_activity_query_size': '16384',
      'idle_in_transaction_session_timeout': '7200000',
    },
  }),
});
globaldbM.rdsCluster.connections.allowDefaultPortFrom(ec2.Peer.ipv4(`${process.env.MYIP}/32`))

const stackS = new Stack(mockApp, 'testing-stackS',{env: envSingapro});
const vpcPublic2 = new ec2.Vpc(stackS,'defaultVpc2',{
  natGateways: 0,
  maxAzs: 3,
  subnetConfiguration: [{
    cidrMask: 26,
    name: 'secondVPC2',
    subnetType: ec2.SubnetType.PUBLIC,
  }],
});
const globaldbS = new GolbalAuroraRDSSlaveInfra(stackS, 'slaveregion',{
  vpc: vpcPublic2,subnetType:ec2.SubnetType.PUBLIC, 
});

stackM.addDependency(stackS)


new CfnOutput(stackM, 'password', { value: globaldbM.rdsPassword });
// add second region cluster
globaldbM.addRegionalCluster(stackM,'addregionalrds',{
  region: 'ap-southeast-1',
  dbSubnetGroupName: globaldbS.dbSubnetGroup.dbSubnetGroupName,
});
```

### To deploy
```bash
cdk deploy
```
### To destroy
```bash
cdk destroy
```

## :clap:  Supporters
[![Stargazers repo roster for @neilkuan/cdk-aurora-globaldatabase](https://reporoster.com/stars/neilkuan/cdk-aurora-globaldatabase)](https://github.com/neilkuan/cdk-aurora-globaldatabase/stargazers)
[![Forkers repo roster for @neilkuan/cdk-aurora-globaldatabase](https://reporoster.com/forks/neilkuan/cdk-aurora-globaldatabase)](https://github.com/neilkuan/cdk-aurora-globaldatabase/network/members)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.115](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.114...v0.1.115) (2021-04-21)

### [0.1.114](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.113...v0.1.114) (2021-04-20)

### [0.1.113](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.112...v0.1.113) (2021-04-19)

### [0.1.112](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.111...v0.1.112) (2021-04-18)

### [0.1.111](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.110...v0.1.111) (2021-04-17)

### [0.1.110](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.109...v0.1.110) (2021-04-16)

### [0.1.109](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.108...v0.1.109) (2021-04-15)

### [0.1.108](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.107...v0.1.108) (2021-04-14)

### [0.1.107](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.106...v0.1.107) (2021-04-13)

### [0.1.106](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.105...v0.1.106) (2021-04-12)

### [0.1.105](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.104...v0.1.105) (2021-04-11)

### [0.1.104](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.103...v0.1.104) (2021-04-10)

### [0.1.103](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.102...v0.1.103) (2021-04-09)

### [0.1.102](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.101...v0.1.102) (2021-04-08)

### [0.1.101](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.100...v0.1.101) (2021-04-07)

### [0.1.100](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.99...v0.1.100) (2021-04-06)

### [0.1.99](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.98...v0.1.99) (2021-04-05)

### [0.1.98](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.96...v0.1.98) (2021-03-29)

### [0.1.97](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.55...v0.1.97) (2021-03-28)

### [0.1.96](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.95...v0.1.96) (2021-03-21)

### [0.1.95](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.94...v0.1.95) (2021-02-18)

### [0.1.94](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.93...v0.1.94) (2021-02-17)

### [0.1.93](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.92...v0.1.93) (2021-02-16)

### [0.1.92](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.91...v0.1.92) (2021-02-15)

### [0.1.91](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.90...v0.1.91) (2021-02-14)

### [0.1.90](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.89...v0.1.90) (2021-02-13)

### [0.1.89](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.88...v0.1.89) (2021-02-12)

### [0.1.88](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.87...v0.1.88) (2021-02-11)

### [0.1.87](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.86...v0.1.87) (2021-02-10)

### [0.1.86](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.85...v0.1.86) (2021-02-09)

### [0.1.85](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.84...v0.1.85) (2021-02-08)

### [0.1.84](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.83...v0.1.84) (2021-02-07)

### [0.1.83](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.82...v0.1.83) (2021-02-06)

### [0.1.82](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.81...v0.1.82) (2021-02-05)

### [0.1.81](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.80...v0.1.81) (2021-02-04)

### [0.1.80](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.79...v0.1.80) (2021-01-18)

### [0.1.79](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.78...v0.1.79) (2021-01-16)

### [0.1.78](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.77...v0.1.78) (2021-01-15)

### [0.1.77](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.76...v0.1.77) (2021-01-14)

### [0.1.76](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.75...v0.1.76) (2021-01-14)

### [0.1.75](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.74...v0.1.75) (2021-01-13)

### [0.1.74](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.73...v0.1.74) (2021-01-12)

### [0.1.73](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.72...v0.1.73) (2021-01-11)

### [0.1.72](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.71...v0.1.72) (2021-01-10)

### [0.1.71](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.70...v0.1.71) (2021-01-09)

### [0.1.70](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.69...v0.1.70) (2021-01-08)

### [0.1.69](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.68...v0.1.69) (2021-01-07)

### [0.1.68](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.67...v0.1.68) (2021-01-06)

### [0.1.67](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.66...v0.1.67) (2021-01-05)

### [0.1.66](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.65...v0.1.66) (2021-01-04)

### [0.1.65](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.64...v0.1.65) (2021-01-03)

### [0.1.64](https://github.com/guan840912/cdk-aurora-globaldatabase/compare/v0.1.63...v0.1.64) (2021-01-02)


test/global-rds.test.ts
import * as ec2 from '@aws-cdk/aws-ec2';
import * as _rds from '@aws-cdk/aws-rds';
import { App, Stack } from '@aws-cdk/core';
import { GolbalAuroraRDSMaster, GolbalAuroraRDSSlaveInfra, InstanceTypeEnum, MySQLtimeZone } from '../src/index';
import '@aws-cdk/assert/jest';

const envTokyo = { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'ap-northeast-1' };
test('test create Matser RDS', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });
  new GolbalAuroraRDSMaster(stack, 'GolbalAuroraRDS');
  expect(stack).toHaveResource('AWS::RDS::DBCluster');
  expect(stack).toHaveResource('AWS::RDS::DBClusterParameterGroup', {
    Family: 'aurora-mysql5.7',
    Parameters: {
      time_zone: 'UTC',
    },
  });
  expect(stack).toHaveResource('AWS::RDS::DBInstance', {
    PubliclyAccessible: false,
  });
});

test('test create Matser Vpc Public', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });
  const VPCPublic = new ec2.Vpc(stack, 'defaultVpc', {
    natGateways: 0,
    maxAzs: 3,
    subnetConfiguration: [{
      cidrMask: 26,
      name: 'PublicVPC',
      subnetType: ec2.SubnetType.PUBLIC,
    }],
  });
  new GolbalAuroraRDSMaster(stack, 'GolbalAuroraRDS', { vpc: VPCPublic });
  expect(stack).toHaveResource('AWS::EC2::Subnet', {
    MapPublicIpOnLaunch: true,
  });
  expect(stack).toHaveResource('AWS::RDS::DBInstance', {
    PubliclyAccessible: true,
  });
});

test('test update Parameter Group', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });;
  new GolbalAuroraRDSMaster(stack, 'GolbalAuroraRDS', {
    parameters: {
      time_zone: MySQLtimeZone.ASIA_TAIPEI,
    },
  });
  expect(stack).toHaveResource('AWS::RDS::DBClusterParameterGroup', {
    Parameters: {
      time_zone: 'Asia/Taipei',
    },
  });
});

test('test update timeZone', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });;
  new GolbalAuroraRDSMaster(stack, 'GolbalAuroraRDS', {
    timeZone: MySQLtimeZone.ASIA_TAIPEI,
  });
  expect(stack).toHaveResource('AWS::RDS::DBClusterParameterGroup', {
    Parameters: {
      time_zone: 'Asia/Taipei',
    },
  });
});

test('test change default dbUserName and default database Name', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });;
  new GolbalAuroraRDSMaster(stack, 'GolbalAuroraRDS', {
    dbUserName: 'superuser',
    defaultDatabaseName: 'superdb',
    rdsPassword: '1qaz2wsx',
  });
  expect(stack).toHaveResource('AWS::RDS::DBCluster', {
    Engine: 'aurora-mysql',
    DatabaseName: 'superdb',
    EngineVersion: '5.7.mysql_aurora.2.07.1',
    MasterUsername: 'superuser',
    MasterUserPassword: '1qaz2wsx',
  });
});

test('test no default rdsPassword', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });;
  new GolbalAuroraRDSMaster(stack, 'GolbalAuroraRDS', {
    defaultDatabaseName: 'superdb',
  });
  expect(stack).toHaveResource('AWS::RDS::DBCluster', {
    Engine: 'aurora-mysql',
    DatabaseName: 'superdb',
    EngineVersion: '5.7.mysql_aurora.2.07.1',
    MasterUsername: {
      'Fn::Join': [
        '',
        [
          '{{resolve:secretsmanager:',
          {
            Ref: 'GolbalAuroraRDSRDSClusterSecret55AE520F',
          },
          ':SecretString:username::}}',
        ],
      ],
    },
    MasterUserPassword: {
      'Fn::Join': [
        '',
        [
          '{{resolve:secretsmanager:',
          {
            Ref: 'GolbalAuroraRDSRDSClusterSecret55AE520F',
          },
          ':SecretString:password::}}',
        ],
      ],
    },
  });
});

test('test create Slave region vpc', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });
  new GolbalAuroraRDSSlaveInfra(stack, 'GolbalAuroraRDS');
  expect(stack).toHaveResource('AWS::EC2::VPC');
});

test('test create Slave region use self vpc', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });
  const rdsVpcSecond = new ec2.Vpc(stack, 'RDSVpcRegionSlave', {
    cidr: '10.109.0.0/16',
    enableDnsHostnames: true,
    enableDnsSupport: true,
    natGateways: 1,
  });
  new GolbalAuroraRDSSlaveInfra(stack, 'GolbalAuroraRDS', {
    vpc: rdsVpcSecond,
  });
  expect(stack).toHaveResource('AWS::EC2::VPC');
});

test('test create Slave region vpc default Private Subnet', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });
  new GolbalAuroraRDSSlaveInfra(stack, 'GolbalAuroraRDS');
  expect(stack).toHaveResource('AWS::RDS::DBSubnetGroup', {
    DBSubnetGroupDescription: 'Private Subnets for database',
  });
});


test('test create Slave region vpc use Public Subnet', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });
  new GolbalAuroraRDSSlaveInfra(stack, 'GolbalAuroraRDS', {
    subnetType: ec2.SubnetType.PUBLIC,
  });
  expect(stack).toHaveResource('AWS::RDS::DBSubnetGroup', {
    DBSubnetGroupDescription: 'Public Subnets for database',
  });
});

test('test create Main region vpc use Postgres cluster', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });
  new GolbalAuroraRDSMaster(stack, 'GolbalAuroraRDS', {
    instanceType: InstanceTypeEnum.R5_LARGE,
    rdsPassword: '1qaz2wsx',
    engineVersion: _rds.DatabaseClusterEngine.auroraPostgres({ version: _rds.AuroraPostgresEngineVersion.VER_11_7 }),
    dbClusterpPG: new _rds.ParameterGroup(stack, 'dbClusterparametergroup', {
      engine: _rds.DatabaseClusterEngine.auroraPostgres({
        version: _rds.AuroraPostgresEngineVersion.VER_11_7,
      }),
      parameters: {
        'rds.force_ssl': '1',
        'timezone': 'UTC+8',
      },
    }),
  });
  expect(stack).toHaveResource('AWS::RDS::DBCluster', {
    Engine: 'aurora-postgresql',
    DatabaseName: 'globaldatabase',
    MasterUsername: 'sysadmin',
    MasterUserPassword: '1qaz2wsx',
  });
  expect(stack).toHaveResource('AWS::RDS::DBClusterParameterGroup', {
    Parameters: {
      'timezone': 'UTC+8',
      'rds.force_ssl': '1',
    },
  });
});

test('test Create Custom Resource', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });
  new GolbalAuroraRDSMaster(stack, 'GolbalAuroraRDS', {
    instanceType: InstanceTypeEnum.R5_LARGE,
    rdsPassword: '1qaz2wsx',
    engineVersion: _rds.DatabaseClusterEngine.auroraPostgres({ version: _rds.AuroraPostgresEngineVersion.VER_11_7 }),
    dbClusterpPG: new _rds.ParameterGroup(stack, 'dbClusterparametergroup', {
      engine: _rds.DatabaseClusterEngine.auroraPostgres({
        version: _rds.AuroraPostgresEngineVersion.VER_11_7,
      }),
    }),
  });
  expect(stack).toHaveResource('Custom::UpgradeGlobalClusterProvider');
});

test('test add Regional Function', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envTokyo });
  const globalmainstack = new GolbalAuroraRDSMaster(stack, 'GolbalAuroraRDS', {
    instanceType: InstanceTypeEnum.R5_LARGE,
    rdsPassword: '1qaz2wsx',
    engineVersion: _rds.DatabaseClusterEngine.auroraPostgres({ version: _rds.AuroraPostgresEngineVersion.VER_11_7 }),
    dbClusterpPG: new _rds.ParameterGroup(stack, 'dbClusterparametergroup', {
      engine: _rds.DatabaseClusterEngine.auroraPostgres({
        version: _rds.AuroraPostgresEngineVersion.VER_11_7,
      }),
    }),
  });

  globalmainstack.addRegionalCluster(stack, 'regional', {
    region: 'ap-southeast-1',
    dbSubnetGroupName: 'mock-db-subnet-group-name',
  });

  expect(stack).toHaveResource('Custom::UpgradeGlobalClusterProvider');
  expect(stack).toHaveResource('Custom::addRegionalClusterProvider');
});

const envErrorRegion = { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'sa-east-1' };
test('test error Region', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envErrorRegion });
  expect(()=>{
    const globalmainstack = new GolbalAuroraRDSMaster(stack, 'GolbalAuroraRDS', {
      instanceType: InstanceTypeEnum.R5_LARGE,
      rdsPassword: '1qaz2wsx',
      engineVersion: _rds.DatabaseClusterEngine.auroraPostgres({ version: _rds.AuroraPostgresEngineVersion.VER_11_7 }),
      dbClusterpPG: new _rds.ParameterGroup(stack, 'dbClusterparametergroup', {
        engine: _rds.DatabaseClusterEngine.auroraPostgres({
          version: _rds.AuroraPostgresEngineVersion.VER_11_7,
        }),
      }),
    });

    globalmainstack.addRegionalCluster(stack, 'regional', {
      region: 'ap-southeast-1',
      dbSubnetGroupName: 'mock-db-subnet-group-name',
    });
  }).toThrowError(/This region sa-east-1 not Support Global RDS !!!/);
});

test('test error Region input addRegional Function', () => {
  const app = new App();
  const stack = new Stack(app, 'testing-stack', { env: envErrorRegion });
  expect(()=>{
    new GolbalAuroraRDSSlaveInfra(stack, 'GolbalAuroraRDS');
    expect(stack).toHaveResource('AWS::EC2::VPC');
  }).toThrowError(/This region sa-east-1 not Support Global RDS !!!/);
});

.projenrc.js
const { AwsCdkConstructLibrary } = require('projen');
const { Automation } = require('projen-automate-it');

const PROJECT_NAME = 'cdk-aurora-globaldatabase';
const PROJECT_DESCRIPTION = 'cdk-aurora-globaldatabase is an AWS CDK construct library that provides Cross Region Create Global Aurora RDS Databases.';
const AUTOMATION_TOKEN = 'AUTOMATION_GITHUB_TOKEN';
const CDK_VERSION = '1.96.0';

const project = new AwsCdkConstructLibrary({
  name: PROJECT_NAME,
  description: PROJECT_DESCRIPTION,
  repository: 'https://github.com/guan840912/cdk-aurora-globaldatabase.git',
  authorName: 'Neil Kuan',
  authorEmail: 'guan840912@gmail.com',
  keywords: ['aws', 'cdk', 'rds', 'global'],
  catalog: {
    twitter: 'neil_kuan',
    announce: false,
  },
  compat: true,
  cdkVersion: CDK_VERSION,
  stability: 'experimental',
  dependabot: false,
  defaultReleaseBranch: 'master',
  cdkDependencies: [
    '@aws-cdk/aws-iam',
    '@aws-cdk/aws-ec2',
    '@aws-cdk/aws-rds',
    '@aws-cdk/aws-lambda',
    '@aws-cdk/aws-logs',
    '@aws-cdk/core',
    '@aws-cdk/custom-resources',
    '@aws-cdk/aws-iam',
  ],
  deps: ['projen-automate-it'],
  bundledDeps: ['projen-automate-it'],
  python: {
    distName: 'cdk-aurora-globaldatabase',
    module: 'cdk_aurora_globaldatabase',
  },
  rebuildBot: false,
});

const automation = new Automation(project, {
  automationToken: AUTOMATION_TOKEN,
});
automation.projenYarnUpgrade();

const common_exclude = ['cdk.out', 'cdk.context.json', 'image', 'yarn-error.log', 'coverage', 'venv'];
const common_exclude_image = ['cdk.out', 'cdk.context.json', 'yarn-error.log', 'coverage', 'venv'];
project.gitignore.exclude(...common_exclude_image);

project.npmignore.exclude(...common_exclude);
project.synth();


custom-resource-handler/index.py
# Dependencies
import boto3
import time

ec2 = boto3.client('rds')
data={}
print("[INFO]", "Initialize function")
def on_event(event, context):
    print(event)
    request_type = event['RequestType']
    if request_type == 'Create':
        return on_create(event)
    if request_type == 'Update':
        return on_update(event)
    if request_type == 'Delete':
        return on_delete(event)
    raise Exception("Invalid request type: %s" % request_type)


def on_create(event):
    props = event["ResourceProperties"]
    GlobalClusterIdentifier = props['GlobalClusterIdentifier']
    SourceDBClusterIdentifier = props['SourceDBClusterIdentifier']
    DeletionProtection = props['DeletionProtection']
    StorageEncrypted = props['StorageEncrypted']
    data ={
      'GlobalClusterIdentifier': GlobalClusterIdentifier,
      'SourceDBClusterIdentifier': SourceDBClusterIdentifier,
      'DeletionProtection': DeletionProtection,
      'StorageEncrypted': StorageEncrypted
    }
    # rds.create_global_cluster(
    # )
    if (data):
      output = {'Status': data}
    else:
      output = {'Status': 'Created'}
    return {'Data': output}


def on_update(event):
    output = {'Status': 'Updated'}
    return {'Data': output}



def on_delete(event):
    output = {'Status': 'Updated'}
    return {"Data": output}


custom-resource-handler/global_index.py
# Dependencies
import boto3
import time

rds = boto3.client('rds')
data={}
print("[INFO]", "Initialize function")
def on_event(event, context):
    print(event)
    request_type = event['RequestType']
    if request_type == 'Create':
        return on_create(event)
    if request_type == 'Update':
        return on_update(event)
    if request_type == 'Delete':
        return on_delete(event)
    raise Exception("Invalid request type: %s" % request_type)


def on_create(event):
    props = event["ResourceProperties"]
    GlobalClusterIdentifier_value = props['GlobalClusterIdentifier']
    SourceDBClusterIdentifier_value = props['SourceDBClusterIdentifier']
    response = rds.create_global_cluster(
    GlobalClusterIdentifier=GlobalClusterIdentifier_value,
    SourceDBClusterIdentifier=SourceDBClusterIdentifier_value)
    data ={
      'GlobalClusterIdentifier': GlobalClusterIdentifier_value,
      'SourceDBClusterIdentifier': SourceDBClusterIdentifier_value,
      'Engine': response['GlobalCluster']['Engine'],
      'EngineVersion': response['GlobalCluster']['EngineVersion'],
      'GlobalClusterArn': response['GlobalCluster']['GlobalClusterArn'],
    }
    if (data):
      output = {'Data': data}
    else:
      output = {'Status': 'Created'}
    return output


def on_update(event):
    print("[INFO]", "Update Event")
    props = event["ResourceProperties"]
    print("update resource with props %s" % (props))
    output = {'Status': 'Updated'}
    return output



def on_delete(event):
    print("[INFO]", "Delete Event")
    props = event["ResourceProperties"]
    GlobalClusterIdentifier_value = props['GlobalClusterIdentifier']
    SourceDBClusterIdentifier_value = props['SourceDBClusterIdentifier']
    time.sleep(100)
    response = rds.remove_from_global_cluster(
    GlobalClusterIdentifier=GlobalClusterIdentifier_value,
    DbClusterIdentifier=SourceDBClusterIdentifier_value)
    print(response)
    time.sleep(5)
    if not response['GlobalCluster']['GlobalClusterMembers']:
       rds.delete_global_cluster(
       GlobalClusterIdentifier=GlobalClusterIdentifier_value)
       removeCluster=True
       time.sleep(5)

    data ={
      'GlobalClusterIdentifier': GlobalClusterIdentifier_value,
      'SourceDBClusterIdentifier': SourceDBClusterIdentifier_value,
      'Status': removeCluster
    }
    if (data):
      output = {'Data': data}
    else:
      output = {'Status': 'deleted'}
    return output

src/integ.default.ts
import { GolbalAuroraRDSMaster, InstanceTypeEnum, GolbalAuroraRDSSlaveInfra } from './index';
import { App, Stack, CfnOutput } from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as _rds from '@aws-cdk/aws-rds';

const mockApp = new App();
const envSingapro  = { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'ap-southeast-1' };
const envTokyo = { account: process.env.CDK_DEFAULT_ACCOUNT, region: 'ap-northeast-1' };

const stackM = new Stack(mockApp, 'testing-stackM',{env: envTokyo});
const vpcPublic = new ec2.Vpc(stackM,'defaultVpc',{
  natGateways: 0,
  maxAzs: 3,
  subnetConfiguration: [{
    cidrMask: 26,
    name: 'masterVPC2',
    subnetType: ec2.SubnetType.PUBLIC,
  }],
});
const globaldbM = new GolbalAuroraRDSMaster(stackM, 'golbalAuroraRDSMaster',{
  instanceType: InstanceTypeEnum.R5_LARGE,
  vpc: vpcPublic,
  //rdsPassword: '1qaz2wsx',
  engineVersion: _rds.DatabaseClusterEngine.auroraPostgres({
    version: _rds.AuroraPostgresEngineVersion.VER_11_7}),
  dbClusterpPG: new _rds.ParameterGroup(stackM, 'dbClusterparametergroup', {
    engine: _rds.DatabaseClusterEngine.auroraPostgres({
      version: _rds.AuroraPostgresEngineVersion.VER_11_7,
    }),
    parameters: {
      'rds.force_ssl': '1',
      'rds.log_retention_period': '10080',
      'auto_explain.log_min_duration': '5000',
      'auto_explain.log_verbose': '1',
      'timezone': 'UTC+8',
      'shared_preload_libraries': 'auto_explain,pg_stat_statements,pg_hint_plan,pgaudit',
      'log_connections': '1',
      'log_statement': 'ddl',
      'log_disconnections': '1',
      'log_lock_waits': '1',
      'log_min_duration_statement': '5000',
      'log_rotation_age': '1440',
      'log_rotation_size': '102400',
      'random_page_cost': '1',
      'track_activity_query_size': '16384',
      'idle_in_transaction_session_timeout': '7200000',
    },
  }),
});
globaldbM.rdsCluster.connections.allowDefaultPortFrom(ec2.Peer.ipv4(`${process.env.MYIP}/32`))

const stackS = new Stack(mockApp, 'testing-stackS',{env: envSingapro});
const vpcPublic2 = new ec2.Vpc(stackS,'defaultVpc2',{
  natGateways: 0,
  maxAzs: 3,
  subnetConfiguration: [{
    cidrMask: 26,
    name: 'secondVPC2',
    subnetType: ec2.SubnetType.PUBLIC,
  }],
});
const globaldbS = new GolbalAuroraRDSSlaveInfra(stackS, 'slaveregion',{
  vpc: vpcPublic2,subnetType:ec2.SubnetType.PUBLIC, 
});

stackM.addDependency(stackS)


new CfnOutput(stackM, 'password', { value: `${globaldbM.rdsPassword}`});
// add second region cluster
globaldbM.addRegionalCluster(stackM,'addregionalrds',{
  region: 'ap-southeast-1',
  dbSubnetGroupName: globaldbS.dbSubnetGroup.dbSubnetGroupName,
});